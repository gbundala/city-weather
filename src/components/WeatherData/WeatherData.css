.weather-data-wrapper img {
  height: 30px;
  width: 30px;
}
