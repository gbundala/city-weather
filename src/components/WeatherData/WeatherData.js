import React from "react";
import ListGroup from "react-bootstrap/ListGroup";
import Card from "react-bootstrap/Card";
import "./WeatherData.css";

// PNG imports for the icons
import humidity from "./images/humidity.png";
import windIcon from "./images/wind-signal.png";
import clouds from "./images/clouds.png";

/**
 *
 *  Weather Data Presentational Component:
 *
 * This is basically a presentational component, that aims
 * to display the results of the data that is being
 * fetched in the CityWeather component. We have therefore
 * separated the presentational part of the app from the
 * business logic part of it
 *
 * We use React Boostrap for the presentational bit and pass
 * the data received through props into this component
 * inside the elements that will be rendered in the DOM
 * as innerHTML of the rendered elements
 */

export default function WeatherData({ weatherData }) {
  // we then destructure the top level properties
  // from the weatherData object

  // we include an early return statement so that
  // it fails gracefully when no props are sent
  // we cannot possible provide the default value
  // of the props, hence this is the best alternative
  if (!weatherData) return <p>There is no data!</p>;

  const { main, weather, wind, name, sys } = weatherData;

  return (
    <div className="weather-data-wrapper">
      <link
        rel="stylesheet"
        href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css"
        integrity="sha384-1BmE4kWBq78iYhFldvKuhfTAU6auU8tT94WrHftjDbrCEXSU1oBoqyl2QvZ6jIW3"
        crossOrigin="anonymous"
      />

      {weather ? (
        <Card style={{ width: "18rem" }} className="container py-5 h-100">
          <Card.Body>
            <Card.Title>{main.temp} °C</Card.Title>
            <Card.Subtitle className="mb-2 text-muted">
              {name} ({sys.country})
            </Card.Subtitle>
            <Card.Text>Feels like: {main.feels_like} °C</Card.Text>
            <Card.Text>
              Max: {main.temp_max}°C ● Min: {main.temp_min}°C
            </Card.Text>
            <ListGroup variant="flush">
              <ListGroup.Item>
                <img src={humidity} alt="humidity" /> {main.humidity}%
              </ListGroup.Item>
              <ListGroup.Item>
                {" "}
                <img src={windIcon} alt="wind" /> {wind.speed} meter/sec
              </ListGroup.Item>
              <ListGroup.Item>
                <img src={clouds} alt="clouds" /> {weather[0].description}
              </ListGroup.Item>
            </ListGroup>
          </Card.Body>
        </Card>
      ) : (
        "Search your city to display the current weather information!"
      )}
    </div>
  );
}
