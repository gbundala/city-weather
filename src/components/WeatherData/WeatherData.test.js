import React from "react";
import renderer from "react-test-renderer";
import WeatherData from "./WeatherData";

// This article has been particularly helpful
// in groking the understanding of snapshot test
// in React applications
// https://www.digitalocean.com/community/tutorials/how-to-write-snapshot-tests-for-react-components-with-jest

// Mocking weatherData
// a sample of weather data to be used in the tests
// for when there is data in the weatherData prop
const weatherData = {
  coord: {
    lon: 39.2695,
    lat: -6.8235,
  },
  weather: [
    {
      id: 802,
      main: "Clouds",
      description: "scattered clouds",
      icon: "03d",
    },
  ],
  base: "stations",
  main: {
    temp: 32.5,
    feels_like: 35.04,
    temp_min: 32.5,
    temp_max: 32.5,
    pressure: 1011,
    humidity: 49,
    sea_level: 1011,
    grnd_level: 1009,
  },
  visibility: 10000,
  wind: {
    speed: 6.64,
    deg: 115,
    gust: 7.03,
  },
  clouds: {
    all: 45,
  },
  dt: 1638008997,
  sys: {
    country: "TZ",
    sunrise: 1637981793,
    sunset: 1638026696,
  },
  timezone: 10800,
  id: 160263,
  name: "Dar es Salaam",
  cod: 200,
};

// snapshot test to match with the existing snapshot
// when there is weatherData
test("renders correctly when there is weatherData", () => {
  const tree = renderer
    .create(<WeatherData weatherData={weatherData} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

// test to match the existing snapshot when there is
// no data
test("renders correctly when there is no weatherData", () => {
  const tree = renderer.create(<WeatherData />).toJSON();

  expect(tree).toMatchSnapshot();
});
