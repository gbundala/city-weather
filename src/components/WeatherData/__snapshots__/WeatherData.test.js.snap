// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when there is no weatherData 1`] = `
<p>
  There is no data!
</p>
`;

exports[`renders correctly when there is weatherData 1`] = `
<div
  className="weather-data-wrapper"
>
  <link
    crossOrigin="anonymous"
    href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css"
    integrity="sha384-1BmE4kWBq78iYhFldvKuhfTAU6auU8tT94WrHftjDbrCEXSU1oBoqyl2QvZ6jIW3"
    rel="stylesheet"
  />
  <div
    className="container py-5 h-100 card"
    style={
      Object {
        "width": "18rem",
      }
    }
  >
    <div
      className="card-body"
    >
      <div
        className="card-title h5"
      >
        32.5
         °C
      </div>
      <div
        className="mb-2 text-muted card-subtitle h6"
      >
        Dar es Salaam
         (
        TZ
        )
      </div>
      <p
        className="card-text"
      >
        Feels like: 
        35.04
         °C
      </p>
      <p
        className="card-text"
      >
        Max: 
        32.5
        °C ● Min: 
        32.5
        °C
      </p>
      <div
        className="list-group list-group-flush"
        onKeyDown={[Function]}
      >
        <div
          className="list-group-item"
          data-rr-ui-event-key={null}
          onClick={[Function]}
        >
          <img
            alt="humidity"
            src="humidity.png"
          />
           
          49
          %
        </div>
        <div
          className="list-group-item"
          data-rr-ui-event-key={null}
          onClick={[Function]}
        >
           
          <img
            alt="wind"
            src="wind-signal.png"
          />
           
          6.64
           meter/sec
        </div>
        <div
          className="list-group-item"
          data-rr-ui-event-key={null}
          onClick={[Function]}
        >
          <img
            alt="clouds"
            src="clouds.png"
          />
           
          scattered clouds
        </div>
      </div>
    </div>
  </div>
</div>
`;
